Add roadmap for 2023 (#70)
in-toto/docs: ROADMAP.md

_Note: Previous roadmaps can be found with the roadmap reviews for that period.
[Link to Roadmap 2022](roadmap-reviews/2022/ROADMAP.md)_

This document spans the Roadmap for the in-toto project for the 2023 calendar
year. As before, the theme of this year's efforts are focused on extending

> in-toto through new and existing ITEs, and adding supporting for them in the

Reviewer: ```suggestion
in-toto through new and existing ITEs, and adding support for them in the
```

implementations we maintain.

> ## in-toto v1.0

Reviewer 2: Good idea, but what does a v1 mean for other versions (e.g., in-toto/attestation)?

Author: Added a note. Does this read okay?

```
Note that this planned v1.0 release checkpoints in-toto at its pre-attestation
stage. As the attestation framework matures and reaches stability, it is likely
to have its own tagged releases, either as a standalone specification or as part
of the main specification's future releases.
```

The last tagged release of the in-toto specification was v0.9. We are gearing
up to tag v1.0 of the specification. This process entails addressing some long
standing discussions and issues, and several rounds of reviews to catch any
inconsistencies and errors that may require breaking changes to fix later.
Apart from these processes, there aren't any major blockers to the

> specification reaching the 1.0 milestone--it has been stable for a significant

Reviewer: ```suggestion
specification reaching the 1.0 milestone -- it has been stable for a significant
```

amount of time and much of in-toto's development happens through in-toto
Enhancements (ITEs). We expect the v1.0 release to happen in the first review
period of this roadmap.

## ITEs

Several important ITEs were introduced in the last few years. ITEs 2 and 3
discussed how to use TUF as a root of trust for in-toto metadata. ITE-4
introduced the notion of non-file artifacts in in-toto, a semantic that is very

> useful in capturing attestations for real world software supply chains. ITE-5

Reviewer 2: Maybe add GitHub PRs as a motivating example for ITE-4?

disassociated the signature wrapper from the in-toto specification and
recommended a switch to DSSE. ITE-6 perhaps had the most impact of all the
ITEs, introducing the in-toto Attestation framework. This was paired with ITE-9
that described a process for the in-toto community to introduce new attestation
types, and established the in-toto attestation maintainers, a group of industry
stakeholders who review and provide feedback about new attestation types. ITE-7

> added support for creating and verifying signatures using X.509 certificates.

Reviewer 2: What happened to ITE-8?

Also, consider adding links to the ITEs?

Author: ITE-8 is mentioned in the section about updated layouts as it needs a bit of a revival once we have policy capabilities for attestations. Added links to other ITEs.

### ITE-6, ITE-9, and Policies for in-toto Attestations

> This year, we are working towards having both of these ITEs accepted. ITE-6 is

Reviewer 2: ```suggestion
This year, we are working towards having both ITE-6 and ITE-9 accepted. ITE-6 is
```

already used in a number of applications, most notably by the SLSA
specification. We believe it's time to accept ITE-6 and continue to develop the
attestation framework. Similarly, ITE-9 is still a draft but the process
described in it is already being applied to several new attestation type
proposals. The attestation maintainers have reviewed proposals for runtime
traces and the Software Supply Chain Attribute Integrity (SCAI) specification.

Besides these existing ITEs, a key aim for this year is to introduce a new ITE

> that proposes modifications to in-toto layouts. As it stands, in-toto layouts

Reviewer 2: Exactly: this is a big open problem IMO.

dictionary". The idea here was to identify how different attestation types
overlap, and when they can be used interchangeably. For example, a SLSA
Provenance attestation contains every information used by in-toto's verification
workflow from in-toto link metadata. We believe that building this out will go a
long way to improving in-toto's usability and a sense of the minimum set of
attestations necessary to secure software supply chains as per current
standards.

### ITE-7

As noted before, this ITE introduced the use of X.509 certificates in in-toto.
This ITE has been implemented in in-toto-golang and other implementations such
as TestifySec's Witness. After another round of reviews at KubeCon late last
year, this ITE is also nearing acceptance.

## in-toto Attestations

Apart from the task of reviewing and providing feedback for the development of
new attestation types, the attestation maintainers also plan to work on updating
and maintaining the in-toto attestation specification itself. The plan is to
focus on the backlog of discussions and issues that have been ongoing in the
last year or so. Further, the maintainers plan to improve the list of reviewed

> predicates.

Reviewer 3: I think one important aspect that we might want to explicitly call out (since we've already been discussing this a bit at the maintainers meetings), is generic or language-independent tooling that the different in-toto implementations can use to generate and verify ITE-6/ITE-9 data. This tooling I think will really help make adoption of ITE-6/ITE-9 much more smoother and streamlined.

> ## in-toto Implementations

Reviewer 2: Maybe mention the work on ITE-2 tooling?

[ITE-7](https://github.com/in-toto/ITE/blob/master/ITE/7) added support for
creating and verifying signatures using X.509 certificates.

### ITE-6, ITE-9, and Policies for in-toto Attestations

This year, we are working towards having ITE-6 and ITE-9 accepted. ITE-6 is
already used in a number of applications, most notably by the SLSA
specification. We believe it's time to accept ITE-6 and continue to develop the
attestation framework. Similarly, ITE-9 is still a draft but the process
described in it is already being applied to several new attestation type
proposals. The attestation maintainers have reviewed proposals for runtime
traces and the Software Supply Chain Attribute Integrity (SCAI) specification.

Besides these existing ITEs, a key aim for this year is to introduce a new ITE
that proposes modifications to in-toto layouts. As it stands, in-toto layouts
cannot be used to verify ITE-6 attestations. With ITE-6 close to being accepted,
the focus is now on closing this gap so that in-toto attestations can be
verified using our verification workflows. Alongside defining new layout
schemas, we hope to revive the work on ITE-8 to introduce a library of in-toto

> policies.

Author: Added timelines for ITEs 6, 7, 8, 9, 10. I think this should address these two comments. Thanks for the review!
